### DISCUSSION TITLE
Bridge setup with TLS & custom CA

### DISCUSSION THREAD

**Discussion by thedmi** [nanomq/nanomq#1624] [Q&A] [answered]
I'm struggling setting up a bridge connection with TLS enabled. The goal is to make NanoMQ connect to an EMQX cloud backend with a custom CA certificate.

I've tried with the following config:

```
bridges.mqtt.emqx1 {
  server = "tls+mqtt-tcp://the-server:8883"

  ssl = { 
    cacertfile = "/etc/certs/ca.pem"
  }
```

But this prints errors:

> 2024-01-23 13:58:25 [1] ERROR /nanomq/nng/src/supplemental/nanolib/conf_ver2.c:382 conf_tls_parse_ver2_base: Read keyfile (null) failed!
>  2024-01-23 13:58:26 [19] ERROR /nanomq/nanomq/apps/broker.c:281 server_cb: RECV nng aio result error: 31 or NULL msg received
>  2024-01-23 13:58:25 [28] WARN  /nanomq/nanomq/bridge.c:958 bridge_tcp_disconnect_cb: bridge client disconnected! RC [136]

After that, NanoMQ enters an infinite reconnect loop.

What is the correct way to use the MQTT-TCP bridge over TLS when a custom CA cert is in place server-side?

**Comment by StargazerWayne** [comment, marked as answer]
Make sure you enable ssl/tls in the server-side correctly. You can check [how](https://www.emqx.io/docs/en/latest/network/emqx-mqtt-tls.html) if needed. If ssl/tls in server-side is enabled, then your config **should** work fine. You may upgrade the version you are using, I see some logs are outdated.

**Reply by thedmi** [reply]
Thanks, it was indeed a server-side configuration error.

The above logs were generated by NanoMQ 0.20.8, which is the latest release according to Github. Why do you think they are outdated?

**Reply by StargazerWayne** [reply]
I guess it's just been too long since the last release XD. There will be a new release this month.

**Comment by JaylinYu** [comment]
Your config is valid. we tested it with a self-established remote broker, and it works.
this is more likely an SSL error, please check the logs on the server side.
FYI, nanomq uses mbedtls, it acts more strictly than OpenSSL, plz verify your cert. and we are not experts on TLS/SSL debugging, cannot solve this issue for you directly.

**Comment by Florida-Rob** [comment]
related, does nanomq support Ubuntu 22.04 with Openssl 3.0.2 ? ( After days of testing i'm frustrated that mosquito seems to be stuck with openssl 3.0.2 on the client cert)

**Reply by JaylinYu** [reply]
As mentioned above, nanomq' tls layer build on the top of mbedtls, Nut sure what feature you are going for of openssl 3, just have a try.

**Comment by Florida-Rob** [comment]
Thanks for the reply.  I missed that nanomq uses mbedtls
Our server is Webmin at its core, and uses OpenSSL 3.0.2 as part of the underlying Ubuntu 22.04, so cannot change this.
tried for days to get mosquitto to work with OpenSSL 3.0.2 and it always rejects client certs created from the same CA - no idea, reading blogs think its a 3.0.2 issue), hence looking for alternative as blocked with mosquitto (and no one replies with any hints, so must be a bad bug)
I don't know if the 2 can live side by side ? and just the MQTT use the mbedtls?
why this has to be so hard....
